File: app/src/main/java/com/pavelprymak/propodcast/services/PlayerService.java
package com.pavelprymak.propodcast.services;

import android.app.Notification;
import android.app.Service;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.IBinder;
import android.os.PowerManager;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.pavelprymak.propodcast.App;
import com.pavelprymak.propodcast.R;
import com.pavelprymak.propodcast.utils.otto.player.EventPlayerError;
import com.pavelprymak.propodcast.utils.otto.player.EventUpdateDurationAndCurrentPos;
import com.pavelprymak.propodcast.utils.otto.player.EventUpdateLoading;
import com.pavelprymak.propodcast.utils.otto.player.EventUpdatePlayPauseBtn;
import com.pavelprymak.propodcast.utils.otto.player.EventUpdatePlayerView;
import com.pavelprymak.propodcast.utils.otto.player.EventUpdateTrackImageAndTitle;
import com.pavelprymak.propodcast.utils.player.PlayerErrorsListener;
import com.pavelprymak.propodcast.utils.player.PlayerHelper;
import com.pavelprymak.propodcast.utils.player.PlayerStateListener;
import com.pavelprymak.propodcast.utils.player.UpdateByTimerHandler;
import com.pavelprymak.propodcast.utils.widget.LastTrackPreferenceManager;
import com.pavelprymak.propodcast.utils.widget.WidgetUpdateManager;
import com.squareup.otto.Bus;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import timber.log.Timber;

import static com.pavelprymak.propodcast.App.CHANNEL_ID;

public class PlayerService extends Service implements PlayerStateListener, PlayerErrorsListener {
    public static final String EXTRA_COMMAND_PLAYER = "extraCommandToPlayerService";
    private static final int NOTIFICATION_FOREGROUND_ID = 123;

    public static final String COMMAND_START_TRACK = "commandStartTrack";
    public static final String COMMAND_CONTINUE_LAST_TRACK = "commandContinueLastTrack";
    public static final String COMMAND_STOP_SERVICE = "commandStopService";
    public static final String EXTRA_TRACK_URL = "extraTrackUrl";
    public static final String EXTRA_TRACK_TITLE = "extraTrackTitle";
    public static final String EXTRA_TRACK_AUTHOR = "extraTrackAuthor";
    public static final String EXTRA_TRACK_IMAGE_URL = "extraTrackImageUrl";

    public static final String COMMAND_PAUSE = "commandPause";

    public static final String COMMAND_PLAY = "commandPlay";
    public static final String COMMAND_SEEK_TO_POSITION = "commandSeekToPosition";
    public static final String EXTRA_TRACK_SEEK_PROGRESS_IN_PERSENTS = "extraTrackSeekPosition";

    public static final String COMMAND_UPDATE_UI = "commandUpdateUI";

    public static boolean isStartService = false;
    public static boolean isTrackPlayNow = false;

    private PlayerHelper mPlayerHelper;
    private PowerManager.WakeLock wakeLock;

    private UpdateByTimerHandler mUpdateUIPositionHandler;
    private final Bus eventBus = App.eventBus;
    private final LastTrackPreferenceManager mLastTrackPreferenceManager = App.mLastTrackSettings;


    @Override
    public void onCreate() {
        super.onCreate();
        isStartService = true;
        Timber.d("onCreate");
        PowerManager powerManager = (PowerManager) getSystemService(POWER_SERVICE);
        wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK,
                "ProPodcastApp:Wakelock");
        wakeLock.acquire();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                    .setContentTitle(getString(R.string.app_name))
                    .setContentText("Running...")
                    .setSmallIcon(R.mipmap.ic_launcher)
                    .build();

            startForeground(NOTIFICATION_FOREGROUND_ID, notification);
        }
        Timber.d("Wakelock acquired");
        mPlayerHelper = new PlayerHelper(getApplicationContext(), this, this);
        mUpdateUIPositionHandler = new UpdateByTimerHandler() {
            @Override
            public void doOperation() {
                if (mPlayerHelper != null) {
                    //Update current position and duration
                    if (eventBus != null)
                        eventBus.post(new EventUpdateDurationAndCurrentPos(mPlayerHelper.getCurrentResumePosition(), mPlayerHelper.getCurrentTrackDuration()));
                }
            }
        };
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mPlayerHelper != null) {
            if (mLastTrackPreferenceManager != null) {
                mLastTrackPreferenceManager.saveTrackCurrentPosition(mPlayerHelper.getCurrentResumePosition());
            }
            mPlayerHelper.releasePlayer();
            mPlayerHelper = null;
        }
        if (mUpdateUIPositionHandler != null) {
            mUpdateUIPositionHandler.stopHandler();
            mUpdateUIPositionHandler = null;
        }
        isStartService = false;
        isTrackPlayNow = false;
        WidgetUpdateManager.updateWidget(getApplicationContext());
        Timber.d("onDestroy");
        if (wakeLock.isHeld()) {
            wakeLock.release();
        }
        Timber.d("Wakelock released");
    }


    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String input = intent.getStringExtra(EXTRA_COMMAND_PLAYER);
        Timber.d("onStartCommand");
        Timber.d(input);
        switch (input) {
            case COMMAND_START_TRACK: {
                String trackTitle = intent.getStringExtra(EXTRA_TRACK_TITLE);
                String trackUrlStr = intent.getStringExtra(EXTRA_TRACK_URL);
                String trackAuthor = intent.getStringExtra(EXTRA_TRACK_AUTHOR);
                String trackImage = intent.getStringExtra(EXTRA_TRACK_IMAGE_URL);
                Uri trackUri = Uri.parse(trackUrlStr);
                startTrack(trackTitle, trackAuthor, trackImage, trackUri, 0);
                mLastTrackPreferenceManager.saveTrackInfo(trackUrlStr, trackTitle, trackAuthor, trackImage);
                break;
            }
            case COMMAND_CONTINUE_LAST_TRACK: {
                if (mLastTrackPreferenceManager != null && mLastTrackPreferenceManager.getTrackAudioUrl() != null) {
                    Uri trackUri = Uri.parse(mLastTrackPreferenceManager.getTrackAudioUrl());
                    startTrack(mLastTrackPreferenceManager.getTrackTitle(),
                            mLastTrackPreferenceManager.getTrackAuthor(),
                            mLastTrackPreferenceManager.getTrackLogo(),
                            trackUri,
                            mLastTrackPreferenceManager.getTrackCurrentPosition());
                }
                break;
            }
            case COMMAND_PAUSE: {
                if (mPlayerHelper != null) {
                    mPlayerHelper.pauseTrack();
                    if (mLastTrackPreferenceManager != null)
                        mLastTrackPreferenceManager.saveTrackCurrentPosition(mPlayerHelper.getCurrentResumePosition());
                }
                if (mUpdateUIPositionHandler != null) {
                    mUpdateUIPositionHandler.stopHandler();
                }

                break;
            }
            case COMMAND_PLAY: {
                if (mPlayerHelper != null) {
                    mPlayerHelper.playTrack();
                }
                if (mUpdateUIPositionHandler != null) {
                    mUpdateUIPositionHandler.startHandler();
                }
                break;
            }
            case COMMAND_SEEK_TO_POSITION: {
                float progressInPercents = intent.getFloatExtra(EXTRA_TRACK_SEEK_PROGRESS_IN_PERSENTS, 0f);
                if (mPlayerHelper != null) {
                    mPlayerHelper.seekToPosition(progressInPercents);
                }
                break;
            }
            case COMMAND_UPDATE_UI: {
                if (mPlayerHelper != null && eventBus != null) {
                    eventBus.post(getDataAboutPlayer());
                }
                break;
            }
            case COMMAND_STOP_SERVICE: {
                stopSelf();
                break;
            }
        }
        return START_STICKY;
    }

    private void startTrack(String trackTitle, String trackAuthor, String trackImage, Uri trackUri, long oldPosition) {
        if (mPlayerHelper != null) {
            //update UI loading flag true
            if (eventBus != null)
                eventBus.post(new EventUpdateLoading(true));
            if (mUpdateUIPositionHandler != null) {
                mUpdateUIPositionHandler.startHandler();
            }
            //clear position data and stop
            mPlayerHelper.stopCurrentTrack();
            mPlayerHelper.clearResumePosition();
            //init new track data
            if (oldPosition == 0L) {
                mPlayerHelper.initializePlayer(trackUri);
            } else {
                mPlayerHelper.initializePlayer(trackUri, oldPosition);
            }
            mPlayerHelper.setTrackTitle(trackTitle);
            mPlayerHelper.setTrackImageUrl(trackImage);
            mPlayerHelper.setTrackAuthor(trackAuthor);
            //update UI track image and title
            if (eventBus != null)
                eventBus.post(new EventUpdateTrackImageAndTitle(trackTitle, trackImage));
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private EventUpdatePlayerView getDataAboutPlayer() {
        EventUpdatePlayerView event = new EventUpdatePlayerView();
        if (mPlayerHelper != null) {
            event.setPlay(mPlayerHelper.isPlayerReadyAndPlay());
            event.setLoad(mPlayerHelper.isPlayerLoading());
            event.setTrackTitle(mPlayerHelper.getTrackTitle());
            event.setTrackImageUrl(mPlayerHelper.getTrackImageUrl());
            event.setTrackDuration(mPlayerHelper.getCurrentTrackDuration());
            event.setTrackCurrentPosition(mPlayerHelper.getCurrentResumePosition());
        }
        return event;
    }

    @Override
    public void stateChanged() {
        // Show foreground notification
        if (mPlayerHelper != null && mPlayerHelper.getMediaSessionHelper() != null) {
            Picasso.get().load(mPlayerHelper.getTrackImageUrl()).into(new Target() {
                @Override
                public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                    Notification notification = mPlayerHelper.getMediaSessionHelper().getNotification(mPlayerHelper.getTrackTitle(), mPlayerHelper.getTrackAuthor(), bitmap);
                    if (notification != null) {
                        startForeground(NOTIFICATION_FOREGROUND_ID, notification);
                    }
                }

                @Override
                public void onBitmapFailed(Exception e, Drawable errorDrawable) {
                    Notification notification = mPlayerHelper.getMediaSessionHelper().getNotification(mPlayerHelper.getTrackTitle(), mPlayerHelper.getTrackAuthor(), BitmapFactory.decodeResource(getApplicationContext().getResources(), R.mipmap.ic_launcher));
                    if (notification != null) {
                        startForeground(NOTIFICATION_FOREGROUND_ID, notification);
                    }
                }

                @Override
                public void onPrepareLoad(Drawable placeHolderDrawable) {

                }
            });
        }
    }

    @Override
    public void isBuffed() {
        if (eventBus != null) {
            eventBus.post(new EventUpdateLoading(true));
        }
    }

    @Override
    public void isReadyAndPlay() {
        if (eventBus != null) {
            eventBus.post(new EventUpdateLoading(false));
            eventBus.post(new EventUpdatePlayPauseBtn(true));
        }
        if (mUpdateUIPositionHandler != null && mUpdateUIPositionHandler.isStopHandler()) {
            mUpdateUIPositionHandler.startHandler();
        }
        isTrackPlayNow = true;
        WidgetUpdateManager.updateWidget(getApplicationContext());
    }

    @Override
    public void isReadyAndPause() {
        if (eventBus != null) {
            eventBus.post(new EventUpdateLoading(false));
            eventBus.post(new EventUpdatePlayPauseBtn(false));
        }
        isTrackPlayNow = false;
        WidgetUpdateManager.updateWidget(getApplicationContext());
    }

    @Override
    public void isEnded() {
        if (eventBus != null) {
            eventBus.post(new EventUpdateLoading(false));
        }
        isTrackPlayNow = false;
        WidgetUpdateManager.updateWidget(getApplicationContext());
    }

    @Override
    public void onPlayerError(int errorCode) {
        if (eventBus != null) {
            eventBus.post(new EventPlayerError(errorCode));
        }
    }
}

File: app/src/main/java/com/pavelprymak/propodcast/services/PlayerService.kt
package com.pavelprymak.propodcast.services

import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import androidx.core.app.NotificationCompat
import com.pavelprymak.propodcast.App
import com.pavelprymak.propodcast.R
import com.pavelprymak.propodcast.utils.otto.player.*
import com.pavelprymak.propodcast.utils.player.PlayerErrorsListener
import com.pavelprymak.propodcast.utils.player.PlayerHelper
import com.pavelprymak.propodcast.utils.player.PlayerStateListener
import com.pavelprymak.propodcast.utils.player.UpdateByTimerHandler
import com.pavelprymak.propodcast.utils.widget.LastTrackPreferenceManager
import com.pavelprymak.propodcast.utils.widget.WidgetUpdateManager
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import org.koin.android.ext.android.inject
import timber.log.Timber

const val EXTRA_COMMAND_PLAYER = "extraCommandToPlayerService"
private const val NOTIFICATION_FOREGROUND_ID = 123

const val COMMAND_START_TRACK = "commandStartTrack"
const val COMMAND_CONTINUE_LAST_TRACK = "commandContinueLastTrack"
const val COMMAND_STOP_SERVICE = "commandStopService"
const val EXTRA_TRACK_URL = "extraTrackUrl"
const val EXTRA_TRACK_TITLE = "extraTrackTitle"
const val EXTRA_TRACK_AUTHOR = "extraTrackAuthor"
const val EXTRA_TRACK_IMAGE_URL = "extraTrackImageUrl"
const val COMMAND_PAUSE = "commandPause"
const val COMMAND_PLAY = "commandPlay"
const val COMMAND_SEEK_TO_POSITION = "commandSeekToPosition"
const val EXTRA_TRACK_SEEK_PROGRESS_IN_PERSENTS = "extraTrackSeekPosition"
const val COMMAND_UPDATE_UI = "commandUpdateUI"

class PlayerService : Service(), PlayerStateListener, PlayerErrorsListener {
    private var mPlayerHelper: PlayerHelper? = null
    private var wakeLock: PowerManager.WakeLock? = null

    private var mUpdateUIPositionHandler: UpdateByTimerHandler? = null
    private val mEventBus = App.eventBus
    private val mLastTrackPref: LastTrackPreferenceManager by inject()

    private val dataAboutPlayer: EventUpdatePlayerView
        get() {
            val event = EventUpdatePlayerView()
            mPlayerHelper?.apply {
                event.isPlay = isPlayerReadyAndPlay
                event.isLoad = isPlayerLoading
                event.trackTitle = trackTitle
                event.trackImageUrl = trackImageUrl
                event.trackDuration = currentTrackDuration
                event.trackCurrentPosition = currentResumePosition
            }
            return event
        }


    override fun onCreate() {
        super.onCreate()
        Timber.d("onCreate")
        val powerManager = getSystemService(Context.POWER_SERVICE) as PowerManager
        wakeLock =
            powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, getString(R.string.player_service_wake_lock_tag))
        wakeLock?.acquire()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notification = NotificationCompat.Builder(this, getString(R.string.player_notification_channel_id))
                .setContentTitle(getString(R.string.app_name))
                .setContentText(getString(R.string.player_service_running))
                .setSmallIcon(R.mipmap.ic_launcher)
                .build()
            startForeground(NOTIFICATION_FOREGROUND_ID, notification)
        }
        Timber.d("Wakelock acquired")
        mPlayerHelper = PlayerHelper(applicationContext, this, this)
        mUpdateUIPositionHandler = object : UpdateByTimerHandler() {
            public override fun doOperation() {
                mPlayerHelper?.apply {
                    //Update current position and duration
                    mEventBus.post(EventUpdateDurationAndCurrentPos(currentResumePosition, currentTrackDuration))
                }
            }
        }
        isStartService = true
    }

    override fun onDestroy() {
        super.onDestroy()
        isStartService = false
        isTrackPlayNow = false
        mPlayerHelper?.let { playerHelper ->
            mLastTrackPref.saveTrackCurrentPosition(playerHelper.currentResumePosition)
            playerHelper.releasePlayer()
            mPlayerHelper = null
        }
        mUpdateUIPositionHandler?.stopHandler()
        mUpdateUIPositionHandler = null
        WidgetUpdateManager.updateWidget(applicationContext)
        Timber.d("onDestroy")
        if (wakeLock?.isHeld == true) {
            wakeLock?.release()
        }
        Timber.d("Wakelock released")
    }


    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val input = intent.getStringExtra(EXTRA_COMMAND_PLAYER)
        Timber.d("onStartCommand")
        Timber.d(input)
        when (input) {
            COMMAND_START_TRACK -> {
                val trackTitle = intent.getStringExtra(EXTRA_TRACK_TITLE)
                val trackUrlStr = intent.getStringExtra(EXTRA_TRACK_URL)
                val trackAuthor = intent.getStringExtra(EXTRA_TRACK_AUTHOR)
                val trackImage = intent.getStringExtra(EXTRA_TRACK_IMAGE_URL)
                if (!trackUrlStr.isNullOrEmpty()) {
                    val trackUri = Uri.parse(trackUrlStr)
                    startTrack(trackTitle, trackAuthor, trackImage, trackUri, 0)
                    mLastTrackPref.saveTrackInfo(trackUrlStr, trackTitle, trackAuthor, trackImage)
                }
            }
            COMMAND_CONTINUE_LAST_TRACK -> {
                if (!mLastTrackPref.trackAudioUrl.isNullOrEmpty()) {
                    val trackUri = Uri.parse(mLastTrackPref.trackAudioUrl)
                    startTrack(
                        mLastTrackPref.trackTitle,
                        mLastTrackPref.trackAuthor,
                        mLastTrackPref.trackLogo,
                        trackUri,
                        mLastTrackPref.trackCurrentPosition
                    )
                }
            }
            COMMAND_PAUSE -> {
                mPlayerHelper?.apply {
                    pauseTrack()
                    mLastTrackPref.saveTrackCurrentPosition(currentResumePosition)
                }
                mUpdateUIPositionHandler?.stopHandler()
            }
            COMMAND_PLAY -> {
                mPlayerHelper?.playTrack()
                mUpdateUIPositionHandler?.startHandler()
            }
            COMMAND_SEEK_TO_POSITION -> {
                val progressInPercents = intent.getFloatExtra(EXTRA_TRACK_SEEK_PROGRESS_IN_PERSENTS, 0f)
                mPlayerHelper?.seekToPosition(progressInPercents)
            }
            COMMAND_UPDATE_UI -> {
                if (mPlayerHelper != null) {
                    mEventBus.post(dataAboutPlayer)
                }
            }
            COMMAND_STOP_SERVICE -> {
                stopSelf()
            }
        }
        return START_STICKY
    }

    private fun startTrack(title: String?, author: String?, image: String?, trackUri: Uri, oldPosition: Long) {
        mPlayerHelper?.apply {
            this.trackTitle = title
            this.trackImageUrl = image
            this.trackAuthor = author
            //clear position data and stop
            stopCurrentTrack()
            clearResumePosition()
            //init new track data
            if (oldPosition == 0L) {
                initializePlayer(trackUri)
            } else {
                initializePlayer(trackUri, oldPosition)
            }

            mUpdateUIPositionHandler?.startHandler()
            //update UI loading flag true
            mEventBus.post(EventUpdateLoading(true))
            //update UI track image and title
            mEventBus.post(EventUpdateTrackImageAndTitle(trackTitle, image))
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun stateChanged() {
        // Show foreground notification
        mPlayerHelper?.apply {
            if (mediaSessionHelper != null && trackImageUrl != null) {
                Picasso.get().load(trackImageUrl).into(object : Target {
                    override fun onPrepareLoad(placeHolderDrawable: Drawable?) {
                    }

                    override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {
                        val notification = mediaSessionHelper
                            .getNotification(
                                trackTitle,
                                trackAuthor,
                                BitmapFactory.decodeResource(applicationContext.resources, R.mipmap.ic_launcher)
                            )
                        if (notification != null) {
                            startForeground(NOTIFICATION_FOREGROUND_ID, notification)
                        }
                    }

                    override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
                        val notification = mediaSessionHelper.getNotification(
                            trackTitle, trackAuthor, bitmap
                        )
                        if (notification != null) {
                            startForeground(NOTIFICATION_FOREGROUND_ID, notification)
                        }
                    }
                })
            }
        }
    }

    override fun isBuffed() {
        mEventBus.post(EventUpdateLoading(true))
    }

    override fun isReadyAndPlay() {
        isTrackPlayNow = true
        mEventBus.post(EventUpdateLoading(false))
        mEventBus.post(EventUpdatePlayPauseBtn(true))
        if (mUpdateUIPositionHandler?.isStopHandler == true) {
            mUpdateUIPositionHandler?.startHandler()
        }
        WidgetUpdateManager.updateWidget(applicationContext)
    }

    override fun isReadyAndPause() {
        isTrackPlayNow = false
        mEventBus.post(EventUpdateLoading(false))
        mEventBus.post(EventUpdatePlayPauseBtn(false))
        WidgetUpdateManager.updateWidget(applicationContext)
    }

    override fun isEnded() {
        isTrackPlayNow = false
        mEventBus.post(EventUpdateLoading(false))
        WidgetUpdateManager.updateWidget(applicationContext)
    }

    override fun onPlayerError(errorCode: Int) {
        mEventBus.post(EventPlayerError(errorCode))
    }

    companion object {
        var isStartService = false
        var isTrackPlayNow = false
    }
}
